components/pci/component.go

```diff
@@ -3,69 +3,258 @@ package pci
 }
 
 func (c *component) Events(ctx context.Context, since time.Time) ([]components.Event, error) {
 	return c.eventBucket.Get(ctx, since)
 }
 
 func (c *component) Metrics(ctx context.Context, since time.Time) ([]components.Metric, error) {
+	log.Logger.Debugw("querying metrics", "since", since)
+
 	return nil, nil
 }
 
 func (c *component) Close() error {
 	log.Logger.Debugw("closing component")
 
-	// safe to call stop multiple times
-	c.poller.Stop(id.Name)
-
+	c.cancel()
 	c.eventBucket.Close()
 
 	return nil
 }
+
+// CheckOnce checks the current pods
+// run this periodically
+func (c *component) CheckOnce() {
+	log.Logger.Infow("checking pci")
+	d := Data{
+		ts: time.Now().UTC(),
+	}
+	defer func() {
+		c.lastMu.Lock()
+		c.lastData = &d
+		c.lastMu.Unlock()
+	}()
+
+	// Virtual machines
+	// Virtual machines require ACS to function, hence disabling ACS is not an option.
+	//
+	// ref. https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/troubleshooting.html#pci-access-control-services-acs
+	if c.currentVirtEnv.IsKVM {
+		return
+	}
+	// unknown virtualization environment
+	if c.currentVirtEnv.Type == "" {
+		return
+	}
+
+	lastEvent, err := c.eventBucket.Latest(c.ctx)
+	if err != nil {
+		d.err = err
+		return
+	}
+
+	nowUTC := time.Now().UTC()
+	if lastEvent != nil && nowUTC.Sub(lastEvent.Time.Time) < 24*time.Hour {
+		log.Logger.Debugw("found events thus skipping -- we only check once per day", "since", humanize.Time(nowUTC))
+		return
+	}
+
+	// in linux, and not in VM
+	// then, check all ACS enabled devices
+	//
+	// Baremetal systems
+	// IO virtualization (also known as VT-d or IOMMU) can interfere with GPU Direct by redirecting all
+	// PCI point-to-point traffic to the CPU root complex, causing a significant performance reduction or even a hang.
+	// If PCI switches have ACS enabled, it needs to be disabled.
+	//
+	// Virtual machines
+	// Virtual machines require ACS to function, hence disabling ACS is not an option.
+	//
+	// ref. https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/troubleshooting.html#pci-access-control-services-acs
+	cctx, cancel := context.WithTimeout(c.ctx, 15*time.Second)
+	d.Devices, err = c.listFunc(cctx)
+	cancel()
+	if err != nil {
+		d.err = err
+		return
+	}
+
+	ev := createEvent(nowUTC, d.Devices)
+	if ev == nil {
+		return
+	}
+
+	// no need to check duplicates
+	// since we check once above
+
+	cctx, cancel = context.WithTimeout(c.ctx, 15*time.Second)
+	err = c.eventBucket.Insert(cctx, *ev)
+	cancel()
+	if err != nil {
+		d.err = err
+		return
+	}
+}
+
+type Data struct {
+	Devices []pci.Device `json:"devices,omitempty"`
+
+	// timestamp of the last check
+	ts time.Time `json:"-"`
+	// error from the last check
+	err error `json:"-"`
+}
+
+func (d *Data) getReason() string {
+	if d == nil {
+		return "no pci data"
+	}
+	if d.err != nil {
+		return fmt.Sprintf("failed to get pci data -- %s", d.err)
+	}
+
+	return fmt.Sprintf("found %d devices", len(d.Devices))
```

Comment:
found % devices, What happened to these devices?? healthy or unhealthy?

Reply:
updated